grpc-java fails to build against protobuf 3.21

<!-- Please answer these questions before submitting a bug report. -->

### What version of gRPC-Java are you using?
1.46.0

### What is your environment?
Linux, Macos, protobuf v3.21

### What did you expect to see?
A successful build

### What did you see instead?
grpc-java fails to build:
```
external/io_grpc_grpc_java/compiler/src/java_plugin/cpp/java_generator.cpp:25:10: fatal error: 'google/protobuf/compiler/java/java_names.h' file not found
#include <google/protobuf/compiler/java/java_names.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

### Steps to reproduce the bug
Attempt to build grpc-java against protobuf 3.21

[This commit](https://github.com/protocolbuffers/protobuf/commit/40db3dabf0fe72b572e89c1e176318b04d3e2408) in upstream `protobuf` renamed `java_names.h` to `names.h`, which is included by `compiler/src/java_plugin/cpp/java_generator.cpp`

--- a/compiler/src/java_plugin/cpp/java_generator.cpp
+++ b/compiler/src/java_plugin/cpp/java_generator.cpp
@@ -22,11 +22,18 @@
 #include <map>
 #include <set>
 #include <vector>
-#include <google/protobuf/compiler/java/java_names.h>
 #include <google/protobuf/descriptor.h>
 #include <google/protobuf/descriptor.pb.h>
 #include <google/protobuf/io/printer.h>
 #include <google/protobuf/io/zero_copy_stream.h>
+#include <google/protobuf/stubs/common.h>
+
+// Protobuf 3.21 changed the name of this file.
+#if GOOGLE_PROTOBUF_VERSION >= 3021000
+  #include <google/protobuf/compiler/java/names.h>
+#else
+  #include <google/protobuf/compiler/java/java_names.h>
+#endif
 
 // Stringify helpers used solely to cast GRPC_VERSION
 #ifndef STR

